File: src/main/java/Java2Kotlin/application/dogs/DogService.java
package Java2Kotlin.application.dogs;

import Java2Kotlin.domain.dog.Dog;
import Java2Kotlin.domain.dog.IDogRepository;
import Java2Kotlin.domain.owner.IOwnerRepository;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class DogService {

    private final IDogRepository dogRepository;

    private final IOwnerRepository ownerRepository;

    public DogService(IDogRepository dogRepository, IOwnerRepository ownerRepository) {
        this.dogRepository = dogRepository;
        this.ownerRepository = ownerRepository;
    }

    public List<Dog> findAll() {
        return dogRepository.all();
    }

    public Dog create(DogUpdateRequest request){
        final Dog dog = new Dog();
        dog.setLocation(request.getLocation());
        dog.setName(request.getName());
        dog.setLastFeed(request.getLastFeed());
        return dogRepository.save(dog);
    }

    public Dog update(String id, DogUpdateRequest request){
        final Dog dog = dogRepository.findById(id);
        if(dog == null){
            throw new DogNotFoundException(id);
        }
        dog.setLocation(request.getLocation());
        dog.setName(request.getName());
        dog.setLastFeed(request.getLastFeed());
        return dogRepository.save(dog);
    }
}

File: src/main/java/java2kotlin/application/dogs/DogService.kt
package java2kotlin.application.dogs

import java2kotlin.domain.dog.Dog
import java2kotlin.domain.dog.IDogRepository
import java2kotlin.domain.owner.IOwnerRepository
import java2kotlin.domain.shared.VaccinationType
import org.springframework.stereotype.Service

@Service
class DogService(
    private val dogRepository: IDogRepository,
    private val ownerRepository: IOwnerRepository
) {

    fun findAll() = dogRepository.all()

    fun create(request: DogUpdateRequest): Dog {
        request.ownerId?.let {
            val foundType = ownerRepository.findById(it)?.address?.country?.requiredVaccinationType
            if(foundType == VaccinationType.TYPE_B){
                throw UnsupportedVaccinationTypeException(foundType)
            }
        }

        val dog = Dog().apply {
            this.location = request.location
            this.name = request.name
            this.lastFeed = request.lastFeed
        }
        return dogRepository.save(dog)
    }

    fun update(id: String?, request: DogUpdateRequest): Dog {
        val dog = dogRepository.findById(id) ?: throw DogNotFoundException(id)
        dog.location = request.location
        dog.name = request.name
        dog.lastFeed = request.lastFeed
        return dogRepository.save(dog)
    }

}
